src/compiler/crystal/types.cr

```diff
@@ -2828,7 +2828,8 @@ module Crystal
     end
 
     def replace_type_parameters(instance)
-      instance_type.replace_type_parameters(instance).metaclass
+      type = instance_type.replace_type_parameters(instance)
+      type == instance_type ? self : type.metaclass
```

Comment:
Seems to be a good candidate for refactoring it into a private method.

Reply:
This is deliberate; what is currently known as `Crystal::MetaclassType` should be a module included by the 4 different metaclass classes in the compiler, but any such changes are orthogonal to this PR, (They might be needed by #11112.)